### PR TITLE
Adding DOFManager from Panzer

### PR DESCRIPTION
**Pull request by gahansen**
This commit is in support of issue #585. It is currently WIP;

    1. Unit tester is "roughed out" but more work is needed to
    plumb it in to test the DOFManager implementation in Albany::BlockedDiscretization.

    2. Some more "plumbing" is needed in the STKConnManager implementation. It does compile
    and most of the internal support for the interface is in there.

### CODE REVIEW

**Comment by gahansen** [comment]
Hi Everyone,

Here is the current state of the DOFManager / STK Connection Manager link into Albany's BlockDiscretization class. There are also unit tests in progress in this PR for these tools.

There is additional discussion of this in #585 . 

The nightly builds will need to include an option for Panzer's DOF manager:

-D Trilinos_ENABLE_PanzerDofMgr:BOOL=ON \

but the rest of Panzer isn't needed.

**Review by lxmota — APPROVED** [review]
I'm out of the loop on this, but I will approve so that I don't hold this up.

File: CMakeLists.txt
Comment (mperego) [inline CMakeLists.txt:304]: I think we no longer need MiniTensor. We should probably remove this

File: src/disc/stk/Albany_GenericSTKMeshStruct.cpp
Comment (mperego) [inline src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: Are we completely getting rid of the rebalance capability? We might need it for GMesh

**Review by mperego — COMMENTED** [review]
@gahansen I had a look at your changes. Sorry it took me so long. 
Overall, it looks fine to me, although I didn't spend enough time on it to understand all the pieces. Thanks for working on this.

There are many places ifdef'd out w/ `#if 0`, and I'm not sure whether those are old pieces of code or something that might become part of the code in the future, if you do not want to remove them maybe some comments would help.

It is not much into Albany tradition, but if you could spend some time documenting the block discretization and the connectivity manager it would be great.

I've seen you had unit tests for dof manager and block discretization which is great.

What's the status of the block discretization and of the PR in general?

File: CMakeLists.txt
Comment (gahansen) [reply CMakeLists.txt:304]: Agreed. I just removed this.

File: src/disc/stk/Albany_GenericSTKMeshStruct.cpp
Comment (gahansen) [reply src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: No problem - I'll put this back if we think we might need it. It might be better to just replace this with the stk_balance capability (in the Trilinos stk package) as we are using some pretty old stuff down in the percept directory of Albany for this.

Please say the word and I'll put this back the way it was.

**Comment by gahansen** [comment]
I'm going to push an update for this in the next day if I can that runs a unit test or two. Irina would like to test that when she returns. I've continued to work on the STKConnManager and the unit tests. I am planning to get this PR in good enough shape to merge it by Sept 30, hopefully I can test the STKConnManager on the Exodus file I've introduced into the unit test in that merge.

Unfortunately, I'm running out of time - I'll probably not get the DOFManager unit test working, or much progress on the overall BlockDiscretization class itself. It is "roughed in" but just barely. The STKConnManager was a bit of a task to get to this point, more than I expected initially.

I plan to clean out those #if's - they are there to just give me some more ideas on tests. I'll put a little more commenting in there to help the next developer as much as I can given the remaining time.

File: src/disc/stk/Albany_GenericSTKMeshStruct.cpp
Comment (mperego) [reply src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: Not sure.. @bartgol what do you think? You are the only one using rebalance for GMesh.

Comment (gahansen) [reply src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: I didn't spend a lot of time looking at this, but there is an integration test for the rebalance in:

stk_integration_tests/stk_balance/IntegrationTestLoadBalance.cpp

Looks pretty interesting at a glance.

Comment (bartgol) [reply src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: I think someone else was using GMesh, I think @JustinClough maybe? Anyhow, I would like to keep GMesh, since it's a very user-friendly meshing tool to create non cartesian geometries (even a simple circle).

If we really want to get rid of the rebalance feature, then we should probably create a mini-executable, to create exodus meshes from GMesh inputs (to be run in serial mode). It shouldn't be complicated, but might take some time. If it's not too much effort, maybe we can just maintain rebalance for the time being?

**Comment by mperego** [comment]
> 
> 
> I'm going to push an update for this in the next day if I can that runs a unit test or two. Irina would like to test that when she returns. I've continued to work on the STKConnManager and the unit tests. I am planning to get this PR in good enough shape to merge it by Sept 30, hopefully I can test the STKConnManager on the Exodus file I've introduced into the unit test in that merge.
> 
> Unfortunately, I'm running out of time - I'll probably not get the DOFManager unit test working, or much progress on the overall BlockDiscretization class itself. It is "roughed in" but just barely. The STKConnManager was a bit of a task to get to this point, more than I expected initially.
> 
> I plan to clean out those #if's - they are there to just give me some more ideas on tests. I'll put a little more commenting in there to help the next developer as much as I can given the remaining time.

Thanks Glen. Sounds good. Let us, and in particular @bartgol  know when the PR is ready for a careful review.

**Comment by gahansen** [comment]
Hi Everyone. As you see above, I just forced an update to my latest code. I still have three tests that do not pass to look at - please let me correct those before you run the tests.

There is a simple serial connection manager unit test. I hope to get parallel working, and a couple more tests in before COB next Wednesday. I'm going to also do some commenting. However, please let me know if you'd like to see me prioritize things differently, as I will not continue this after Sept 30 given current plans and funding and I would like to leave this the best way I can given the remaining time.

File: CMakeLists.txt
Comment (gahansen) [reply CMakeLists.txt]: This is gone.

File: src/disc/stk/Albany_GenericSTKMeshStruct.cpp
Comment (gahansen) [reply src/disc/stk/Albany_GenericSTKMeshStruct.cpp:446]: I put this back in the latest push.

**Comment by gahansen** [comment]
I just pushed my latest changes to the PR, the code compiles and the tests run at this point. I'm slowly paring down the list of remaining issues. I'll work on getting the STKConnManager working in parallel and a parallel unit test working next.

File: src/disc/BlockedDiscretization.hpp
Comment (bartgol) [inline src/disc/BlockedDiscretization.hpp:409]: These are ok now that we have a single block. Later on we should:
1. check that the input vector is a Thyra::ProductVector
2. check that the input blk vector and m_blocks have the same number of blocks
3. loop over all blocks, and get call the corresponding method in m_blocks[i] passing only the i-th block of the input vector.

File: src/disc/stk/Albany_AbstractSTKFieldContainer.hpp
Comment (bartgol) [inline src/disc/stk/Albany_AbstractSTKFieldContainer.hpp:29]: Why this undef?

File: src/disc/stk/Albany_STK3DPointStruct.cpp
Comment (bartgol) [inline src/disc/stk/Albany_STK3DPointStruct.cpp:18]: I don't understand why the topology would be PARTICLE. Isn't particle a 0-d topology?

File: src/disc/stk/Albany_STKDiscretization.cpp
Comment (bartgol) [inline src/disc/stk/Albany_STKDiscretization.cpp:819]: Is the code in the `#if 0` supposed to be used later? 

Also, who uses the `addCellField` and `addSolutionField` methods?

Comment (bartgol) [reply src/disc/stk/Albany_STKDiscretization.cpp:819]: Nvm, I found it in the STKConnManager.

File: src/disc/stk/Albany_STK3DPointStruct.cpp
Comment (gahansen) [reply src/disc/stk/Albany_STK3DPointStruct.cpp:18]: Interesting question, I do not know the answer. I think @ikalash might have been the last to touch this operation in the code before I did. Do you know Irina?

I made the change here to combine the part declaration and topology declaration (on line 22 of the main branch) in a single call instead of using two stk::mesh calls as was past practice. This change made it easier for me to pass this information to the STKConnManager (and it simplifies/robustifies these operations in the code).

File: src/disc/stk/Albany_AbstractSTKFieldContainer.hpp
Comment (gahansen) [reply src/disc/stk/Albany_AbstractSTKFieldContainer.hpp:29]: This will go away as soon as I get the parallel unit test for STKConnManager operational. I am trying to track down a memory issue that appears to be tied to the proc_rank field, using this construct.

File: src/disc/BlockedDiscretization.hpp
Comment (gahansen) [reply src/disc/BlockedDiscretization.hpp:409]: Agreed,

File: src/disc/stk/Albany_STKDiscretization.cpp
Comment (gahansen) [reply src/disc/stk/Albany_STKDiscretization.cpp:819]: 👍

**Comment by ikalash** [comment]
Just FYI - I am testing this now.  Please stay tuned.

**Comment by ikalash** [comment]
I tested this in some of my nightlies.  I had to fix a compilation error in CISM-Albany.  @mperego , it might be worth building/running MPAS-Albany with this PR to make sure there are no issues.  I am OK with this getting merged in.  We will need to make sure we add the required configure flag to all the nightlies so the code compiles once this PR is merged in.

**Review by ikalash — APPROVED** [review]
I tested this in some of my nightlies. I had to fix a compilation error in CISM-Albany. @mperego , it might be worth building/running MPAS-Albany with this PR to make sure there are no issues. I am OK with this getting merged in. We will need to make sure we add the required configure flag to all the nightlies so the code compiles once this PR is merged in.

**Comment by gahansen** [comment]
Hi Irina, Thanks for fixing that compilation issue. I had intended to set up a compile of CISM so I could check those but I never did. I've made some more local changes, pretty minor but I did add quite a bit more to the unit test and cleaned/commented in there. I'll put this all together, squash, and force push the result back so @mperego can test with the final if that is OK with everyone.

**Comment by mperego** [comment]
I'm also in favor of merging this. 
@ikalash  I've seen that the fix you did for CISM is already implemented in the mpas case.
Also, it's a good opportunity to test the MALI nightly builds ;-)

**Comment by gahansen** [comment]
Hi All,  Just pushed final updates and Irina's fix. All the tests pass for me on CEE. I'll merge to master once you give the word.

**Comment by mperego** [comment]
I think you can go ahead and merge. I can check MALI after you merge. :-)

**Comment by gahansen** [comment]
Did it. Thanks everyone!!!

**Pull request merged**
